src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala

```diff
@@ -405,24 +403,22 @@ trait ContextErrors {
             }
             val altStr: String = {
               val maxSuggestions = 4
-              if (companionSymbol != NoSymbol) ""
-              else
+              orEmpty(companionSymbol == NoSymbol) {
                 alternatives match {
                   case Nil => ""
-                  case xs  => "\ndid you mean " + StringUtil.oxford(xs.sorted.take(maxSuggestions), "or") + "?"
+                  case xs  => s"did you mean ${StringUtil.oxford(xs.sorted.take(maxSuggestions), "or")}?"
                 }
+              }
             }
-            val semicolon = (
-              if (linePrecedes(qual, sel))
-                "\npossible cause: maybe a semicolon is missing before `"+nameString+"`?"
-              else
-                ""
-            )
-            val notAnyRef = (
-              if (ObjectClass.info.member(name).exists) notAnyRefMessage(target)
-              else ""
-            )
-            companion + altStr + notAnyRef + semicolon
+            val semicolon = orEmpty(linePrecedes(qual, sel))(s"possible cause: maybe a semicolon is missing before `$nameString`?")
+            val notAnyRef = orEmpty(ObjectClass.info.member(name).exists)(notAnyRefMessage(target))
+            val javaRules = orEmpty(owner.isJavaDefined) {
+              def more = orEmpty(settings.explaintypes)(sm""" Static Java members belong to companion objects in Scala;
+                |they are not inherited, even by subclasses defined in Java.""")
+              val (jtype, jmember) = cx.javaFindMember(target, name, _ => true)
```

Comment:
Double check if this is prone to false positive, if it should check for static only, etc. Counterexamples in tests?